packages/shorebird_cli/lib/src/ios_deploy.dart

```diff
@@ -24,25 +44,265 @@ class IOSDeploy {
+          lldbRun = RegExp(RegExp.escape(prompt) + r'\s*run');
+          logger.detail(line);
+          return;
+        }
+
+        // lldb is launching the debugger.
+        // Example: (lldb)     run
+        if (lldbRun.hasMatch(line)) {
+          logger.detail(line);
+          debuggerState = _IOSDeployDebuggerState.launching;
+          return;
+        }
+
+        // The next line after "run" must either be "success"
+        // or attaching the debugger was unsuccessful.
+        if (debuggerState == _IOSDeployDebuggerState.launching) {
+          logger.detail(line);
+          final attachSuccess = line == 'success';
+          debuggerState = attachSuccess
+              ? _IOSDeployDebuggerState.attached
+              : _IOSDeployDebuggerState.detached;
+          launchProgress.complete(
+            attachSuccess ? 'Started app' : 'Failed to start app',
+          );
+          return;
+        }
+
+        // The app has been stopped.
+        if (line.contains('PROCESS_STOPPED') ||
+            _lldbProcessStopped.hasMatch(line)) {
+          logger.detail(line);
+          launchProcess.stdin.writeln(_backTraceAll);
+          detach();
+          return;
+        }
+
+        // The app exited or crashed, so exit. Continue passing debugging
+        // messages to the log reader until it exits to capture crash dumps.
+        if (line.contains('PROCESS_EXITED') ||
+            _lldbProcessExit.hasMatch(line)) {
+          logger.detail(line);
+          kill();
+          return;
+        }
+
+        // The debugger has detached from the app.
+        if (_lldbProcessDetached.hasMatch(line)) {
+          kill();
+          return;
+        }
+
+        // The debugger is resuming.
+        if (_lldbProcessResuming.hasMatch(line)) {
+          logger.detail(line);
+          debuggerState = _IOSDeployDebuggerState.attached;
+          return;
+        }
+
+        // Format progress logs for the user while the debugger is attaching.
+        if (debuggerState != _IOSDeployDebuggerState.attached) {
+          if (line.contains('Copying') && line.endsWith('to device')) {
+            final abbreviatedLine = line.replaceFirst('$bundlePath/', '');
+            launchProgress.update(abbreviatedLine);
+          } else if (line.startsWith(RegExp(r'\[\s+\d+\%\]'))) {
+            launchProgress.update(line);
+          } else {
+            logger.detail(line);
+          }
+          return;
+        }
+
+        // Avoid logging empty lines.
+        if ((previousLine?.isNotEmpty ?? false) && line.isEmpty) return;
+
+        // Output logs after the debugger has attached.
+        if (debuggerState == _IOSDeployDebuggerState.attached) {
+          logger.info(line);
+        }
+
+        previousLine = line;
+      }
+
+      void onStderr(String line) {
+        monitorIOSDeployFailure(line, logger);
+        logger.detail(line);
+      }
+
+      final stdoutSubscription = launchProcess.stdout
```

Comment:
subscribeToLine(launchProcess.stdout, onStdOut) or similar?  Not important, just a little surprised you had to write this yourself.

Reply:
Yeah good call let me create a helper for this to at least simplify things a bit more.